sys/userlib/src/lib.rs

```diff
@@ -1284,71 +1284,161 @@ pub unsafe extern "C" fn _start() -> ! {
+    // RAM):
     const BUFSIZE: usize = 128;
 
-    // If the panic message is longer than BUFSIZE, we'll throw away remaining
-    // bytes. So we need to implement `core::fmt::Write` (how panic messages get
-    // generated) in a way that will only record a maximum-length _prefix_ of
-    // the output.
+    // Panic messages get constructed using `core::fmt::Write`. If we implement
+    // that trait, we can provide our own type that will back the
+    // `core::fmt::Formatter` handed into any formatting routines (like those on
+    // `Debug`/`Display`). This is important, because the standard library does
+    // not provide an implementation of `core::fmt::Write` that uses a fixed
+    // size buffer and discards the rest -- which is what we want.
     //
-    // To do that we have this custom type:
-    struct PrefixWrite([u8; BUFSIZE], usize);
+    // And so, we provide our own!
+    use core::fmt::Write;
+
+    struct PrefixWrite {
+        /// Content will be written here. While the content itself will be
+        /// UTF-8, it may end in an incomplete UTF-8 character to simplify our
+        /// truncation logic.
+        buf: &'static mut [u8; BUFSIZE],
+        /// Number of bytes of `buf` that are valid.
+        ///
+        /// Invariant: always in the range `0..buf.len()`.
+        pos: usize,
+    }
 
     impl Write for PrefixWrite {
         fn write_str(&mut self, s: &str) -> core::fmt::Result {
-            if self.1 < self.0.len() {
-                let (_, remaining) = self.0.split_at_mut(self.1);
-                let strbytes = s.as_bytes();
-                let to_write = remaining.len().min(strbytes.len());
-                unsafe {
-                    core::ptr::copy_nonoverlapping(
-                        strbytes.as_ptr(),
-                        remaining.as_mut_ptr(),
-                        to_write,
-                    );
-                }
-                // Wrapping add here because (1) we're confident that adding
-                // to_write won't cause wrap because of our `min` above, and (2)
-                // we want to avoid writing code that implies an integer
-                // overflow check.
-                self.1 = self.1.wrapping_add(to_write);
+            // If we've filled the buffer, this is a no-op.
+            //
+            // Note that we check `>=` here even though our invariant on this
+            // type states that the `>` part won't happen. This is to reassure
+            // the compiler, which otherwise can't see our invariant.
+            if self.pos >= self.buf.len() {
+                return Ok(());
             }
+
+            // We can write more. How much more? Let's find out. We're going to
+            // do this using unchecked operations to avoid a recursive panic.
+            //
+            // Safety: this is unsafe because of the risk of passing in an
+            // out-of-bounds index for the split. However, our type invariant
```

Comment:
is it the "type invariant" or the check that we made above on 1343? the type of pos is usize, which could hold a value bigger than buf.len()

Reply:
Ah, sry, by "our type invariant" I mean the invariant on PrefixWrite (our type). I could clarify the comments.

Reply:
nah i think you're totally good to merge, hence the approval.